README.md
# Scrollbox

[![Build Status](https://travis-ci.org/Invis1ble/scrollbox.svg?branch=master)](https://travis-ci.org/Invis1ble/scrollbox)
[![Code Climate](https://codeclimate.com/github/Invis1ble/scrollbox/badges/gpa.svg)](https://codeclimate.com/github/Invis1ble/scrollbox)
![Bower version](https://img.shields.io/bower/v/scrollbox.svg)
[![npm version](https://img.shields.io/npm/v/scrollbox.svg)](https://www.npmjs.com/package/scrollbox)
[![Packagist Prerelease](https://img.shields.io/packagist/v/Invis1ble/scrollbox.svg)](https://packagist.org/packages/Invis1ble/scrollbox)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](./LICENSE)

A lightweight jQuery custom scrollbar plugin, that triggers event when reached the defined point.

[Demo Page](https://invis1ble.github.io/scrollbox/)

## Table of contents

- [Browser compatibility](#browser-compatibility)
- [Installation](#installation)
- [Usage](#usage)
- [Basic features](#basic-features)
    - [Options](#options)
    - [Methods](#methods)
    - [Events](#events)
- [Infinite scrolling implementation example](#infinite-scrolling-implementation-example)
- [License](#license)

## Browser compatibility

* IE 7+
* Firefox
* Opera (old and new)
* Safari
* Chrome
* Chrome for Android

... and others.

Scrollbox is automatically tested on the following browsers

[![Sauce Test Status](https://saucelabs.com/browser-matrix/Invis1ble.svg)](https://saucelabs.com/u/Invis1ble)

## Installation

Several quick start options are available:

* [Download the latest release](https://github.com/Invis1ble/scrollbox/archive/master.zip).
* Clone the repo: `git clone https://github.com/Invis1ble/scrollbox.git`.
* Install with [Bower](http://bower.io): `bower install scrollbox`.
* Install with [npm](https://www.npmjs.com/): `npm install scrollbox`.
* Install with [Composer](https://getcomposer.org): `composer require invis1ble/scrollbox`.

After installing the plugin you have to install [jquery-mousewheel](https://github.com/jquery/jquery-mousewheel).
You can simply [download it as archive](https://github.com/jquery/jquery-mousewheel/archive/master.zip) and unpack to desired location.

## Usage

In order to use the plugin, you have to include styles and script to your html e.g.:

```html
<link href="/path/to/scrollbox.min.css" media="screen" rel="stylesheet" type="text/css">
<script src="//ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
<script src="/path/to/jquery.mousewheel.min.js" type="text/javascript"></script>
<script src="/path/to/scrollbox.min.js" type="text/javascript"></script>
```

Let's assume that you want to stylize the following element:

```html
<div id="long-content-container">
    Here is a long content
</div>
```

Then you should define `max-height` and/or `max-width` of that element:

```css
#long-content-container {
    max-height: 200px;
    max-width: 200px;
}
```

and initialize Scrollbox:

```js
$('#long-content-container').scrollbox();
```

That's all.
Now if real size of the container is greater than the specified `max-height`/`max-width` then the plugin will add scrollbar to it.

You can stylize scrollbar via `css` or even better by overwriting corresponding `less` variables.
See [src/less/](https://github.com/Invis1ble/scrollbox/tree/master/src/less) for more details.

## Basic features

### Options

Scrollbox uses several options to configuring behavior. The default options are:

```js
{
    distanceToReach: {
        x: 0,
        y: 0
    },
    wheelSensitivity: 20,
    momentum: {
        acceleration: 1600,
        thresholdTime: 500
    },
    startAt: {
        x: 'left',
        y: 'top'
    },
    templates: {
        horizontalBar: '<div></div>',
        verticalBar: '<div></div>',
        horizontalRail: '<div></div>',
        verticalRail: '<div></div>',
        wrapper: '<div></div>'
    }
}
```

You can optionally pass an object containing all of the options that you would like to initialize Scrollbox with e.g.:

```js
$('#long-content-container').scrollbox({
    wheelSensitivity: 25,
    startAt: {
        y: 'bottom'
    }
});
```

or re-define default values for all instances:

```js
$.fn.scrollbox.Constructor.getDefault().distanceToReach.y = 100;
```

#### startAt.x

The initial position of the scroll on x-axis.

The value can be `'left'`, `'right'` or number of pixels from the left boundary.

#### startAt.y

The initial position of the scroll on y-axis.

The value can be `'top'`, `'bottom'` or number of pixels from the top boundary.

#### distanceToReach.x

The distance from the left and right boundaries of the content when `reachleft.scrollbox` and `reachright.scrollbox` events should be triggered.

This option is useful when you want to implement so-called "infinite scrolling".

#### distanceToReach.y

The distance from the top and bottom boundaries of the content when `reachleft.scrollbox` and `reachright.scrollbox` events should be triggered.

This option is useful when you want to implement so-called "infinite scrolling".

#### wheelSensitivity

The distance in pixels for one fixed step of mouse wheel.

You probably shouldn't change this value.

#### momentum.acceleration

Swipe acceleration factor.

#### momentum.thresholdTime

Threshold time in milliseconds for detect inertial moving at swipe.

#### templates

Normally you don't need to change this templates, but you can if you want.

### Methods

You can call some methods of the plugin.

#### .update()

Recalculates scrollbars' positions and sizes.

For example, If you write the infinite scroll implementation you have to update scrollbar position and size after content has been added.
To do this you should simply call `.update()` method:

```js
$('#long-content-container').scrollbox('update');
```

#### .scrollBy(deltaX, deltaY, animationOptions)

Scrolls by pixels.

See [.animate()](https://api.jquery.com/animate/#animate-properties-options) for the available values of the `animationOptions`

Example:

```js
$('#long-content-container').scrollbox('scrollBy', 100, 200);
```

If you want to scroll only on y-axis you can pass `0` as `deltaX` value:

```js
$('#long-content-container').scrollbox('scrollBy', 0, 200);
```

#### .scrollTo(x, y, animationOptions)

Scrolls to specified position.

`x` can be integer (pixels), or string `'left'` or `'right'`.

`y` also can be integer (pixels), or string `'top'` or `'bottom'`.
 
`'left'`, `'right'`, `'top'` and `'bottom'` means the boundaries.

See [.animate()](https://api.jquery.com/animate/#animate-properties-options) for the available values of the `animationOptions`

Example:

```js
$('#long-content-container').scrollbox('scrollTo', 100, 200);
```

If you want to scroll only on y-axis you can pass `undefined` as `x` value:

```js
$('#long-content-container').scrollbox('scrollTo', undefined, 'bottom');
```

#### .destroy()

Completely removes all stuff from the element.

### Events

Scrollbox triggers a several events during lifecycle.

#### reachleft.scrollbox

Triggered when scrolling reach the left boundary of the content. Respects `distanceToReach.x` option.

#### reachright.scrollbox

Triggered when scrolling reach the right boundary of the content. Respects `distanceToReach.x` option.

#### reachtop.scrollbox

Triggered when scrolling reach the top boundary of the content. Respects `distanceToReach.y` option.

#### reachbottom.scrollbox

Triggered when scrolling reach the bottom boundary of the content. Respects `distanceToReach.y` option.

## Infinite scrolling implementation example

```js
var $container = $('#content-container');

$container
    .on('reachbottom.scrollbox', function () {
        $.ajax({
            // options
        }).done(function (response) {
            $container
                .append(response)
                .scrollbox('update');
        });
    })
    .scrollbox({
        distanceToReach: {
            y: 100
        }
    });
```

## License

[The MIT License](./LICENSE)

CHANGELOG.md
# [v4.0.0](https://github.com/Invis1ble/scrollbox/compare/v3.0.1...v4.0.0)

### Features

* Added table of contents to README.
* Created CHANGELOG.

### Bug Fixes

* Fixed error in IE 7-8.

### Breaking changes from v3

* Default settings are now available via `$.fn.scrollbox.Constructor.getDefault()` instead of `$.fn.scrollbox.Constructor.Default`.


tests/unit/setup.js
(function (QUnit, $) {
    QUnit.config.requireExpects = true;
    QUnit.config.testTimeout = 20000;

    // Display fixture on-screen to avoid false positives
    QUnit.begin(function () {
        $('#qunit-fixture').css({
            top: 0,
            left: 0
        });
    });

    QUnit.done(function () {
        $('#qunit-fixture').css({
            top: '',
            left: ''
        });
    });

    QUnit.testDone(function () {
        $('#qunit-fixture').empty();
    });

    var log = [];
    var testName;

    QUnit.done(function (test_results) {
        var tests = [];

        for (var i = 0, len = log.length; i < len; i ++) {
            var details = log[i];
            tests.push({
                name: details.name,
                result: details.result,
                expected: details.expected,
                actual: details.actual,
                source: details.source
            });
        }

        test_results.tests = tests;

        window.global_test_results = test_results;
    });
    QUnit.testStart(function (testDetails) {
        QUnit.log(function (details) {
            if (!details.result) {
                details.name = testDetails.name;
                log.push(details);
            }
        });
    });
})(QUnit, jQuery);

tests/unit/src/js/setup.js
((QUnit, $) => {
    'use strict';

    QUnit.config.requireExpects = true;
    QUnit.config.testTimeout = 10000;

    // Display fixture on-screen to avoid false positives
    QUnit.begin(() => {
        $('#qunit-fixture').css({
            top: 0,
            left: 0
        });
    });

    QUnit.done(() => {
        $('#qunit-fixture').css({
            top: '',
            left: ''
        });
    });

    QUnit.testDone(() => {
        const $fixture = $('#qunit-fixture');
        const $scrollbox = $('#scrollbox-container', $fixture);

        if (undefined !== $scrollbox.data('scrollbox')) {
            $scrollbox.scrollbox('destroy');
        }

        $fixture.empty();
    });

    let log = [];

    QUnit.done((results) => {
        let tests = [];

        for (let i = 0, len = log.length; i < len; i ++) {
            let details = log[i];

            tests.push({
                name: details.name,
                result: details.result,
                expected: details.expected,
                actual: details.actual,
                source: details.source
            });
        }

        results.tests = tests;

        window.global_test_results = results;
    });

    QUnit.testStart((testDetails) => {
        QUnit.log((details) => {
            if (!details.result) {
                details.name = testDetails.name;
                log.push(details);
            }
        });
    });
})(QUnit, jQuery);

Gruntfile.js
module.exports = (grunt) => {
    'use strict';

    require('load-grunt-tasks')(grunt);

    let testBuildNumber;
    
    if (process.env.TRAVIS_JOB_ID) {
        testBuildNumber = "travis-" + process.env.TRAVIS_JOB_ID;
    } else {
        testBuildNumber = "manual-" + (new Date()).getTime();
    }
    
    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        clean: [
            'dist/*'
        ],
        eslint: {
            src: ['src/js/scrollbox.js']
        },
        babel: {
            options: {
                presets: [
                    ['es2015', { modules: false, loose: true }]
                ]
            },
            production: {
                files: {
                    'dist/js/scrollbox.js': 'src/js/scrollbox.js'
                }
            },
            development: {
                options: {
                    sourceMap: true
                },
                files: {
                    'dist/js/scrollbox.js': 'src/js/scrollbox.js'
                }
            },
            visualTests: {
                options: {
                    sourceMap: true
                },
                files: {
                    'tests/visual/dist/js/setup.js': 'tests/visual/src/js/setup.js'
                }
            },
            unitTests: {
                options: {
                    sourceMap: true
                },
                files: {
                    'tests/unit/dist/js/setup.js': 'tests/unit/src/js/setup.js',
                    'tests/unit/dist/js/scrollbox.js': 'tests/unit/src/js/scrollbox.js'
                }
            }
        },
        uglify: {
            options: {
                report: 'gzip',
                sourceMap: true,
                screwIE8: false,
                banner: `/*!
* Scrollbox v<%= pkg.version %>
* (c) 2013-<%= grunt.template.today('yyyy') %>, <%= pkg.author %>
* Licensed under MIT (https://opensource.org/licenses/mit-license.php)
*/`,
                compress: {
                    unsafe: true
                }
            },
            dist: {
                files: {
                    'dist/js/<%= pkg.name %>.min.js': 'dist/js/<%= pkg.name %>.js'
                }
            }
        },
        less: {
            options: {
                sourceMapRootpath: '/'
            },
            production: {
                files: {
                    'dist/css/<%= pkg.name %>.css': 'src/less/scrollbox.less'
                }
            },
            development: {
                options: {
                    sourceMap: true
                },
                files: {
                    'dist/css/<%= pkg.name %>.css': 'src/less/scrollbox.less'
                }
            },
            visualTests: {
                options: {
                    sourceMap: true
                },
                files: {
                    'tests/visual/dist/css/styles.css': 'tests/visual/src/less/styles.less'
                }
            },
            unitTests: {
                options: {
                    sourceMap: true
                },
                files: {
                    'tests/unit/dist/css/styles.css': 'tests/unit/src/less/styles.less'
                }
            }
        },
        postcss: {
            options: {
                failOnError: true,
                processors: [
                    require('postcss-will-change'),
                    require('postcss-opacity')({
                        legacy: true
                    }),
                    require('autoprefixer')({ browsers: [
                        'last 100 versions', // LOL
                        'ie >= 7'
                    ] })
                ]
            },
            dist: {
                src: 'dist/css/*.css'
            }
        },
        cssmin: {
            options: {
                shorthandCompacting: false,
                roundingPrecision: -1,
                sourceMap: true,
                root: 'dist/css/'
            },
            dist: {
                files: {
                    'dist/css/<%= pkg.name %>.min.css': 'dist/css/<%= pkg.name %>.css'
                }
            }
        },
        modernizr: {
            tests: {
                crawl: false,
                customTests: [],
                dest: 'tests/vendor/modernizr/modernizr-output.js',
                tests: [
                    'touchevents'
                ],
                options: [],
                uglify: false
            }
        },
        qunit: {
            dist: 'tests/unit/index.html'
        },
        watch: {
            srcJs: {
                files: 'src/js/scrollbox.js',
                tasks: ['babel:development']
            },
            srcLess: {
                files: 'src/less/**/*.less',
                tasks: ['less:development']
            },
            visualTestsJs: {
                files: 'tests/visual/src/js/setup.js',
                tasks: ['babel:visualTests']
            },
            visualTestsLess: {
                files: 'tests/visual/src/less/styles.less',
                tasks: ['less:visualTests']
            },
            unitTestsJs: {
                files: 'tests/unit/src/js/*.js',
                tasks: ['babel:unitTests']
            },
            unitTestsLess: {
                files: 'tests/unit/src/less/styles.less',
                tasks: ['less:unitTests']
            }
        },
        connect: {
            tests: {
                options: {
                    base: '.',
                    hostname: '127.0.0.1',
                    port: 9999
                }
            }
        },
        'saucelabs-qunit': {
            all: {
                options: {
                    build: testBuildNumber,
                    throttled: 5,
                    urls: ['http://localhost:9999/tests/unit/index.html'],
                    testname: 'QUnit test for Scrollbox',
                    browsers: grunt.file.readYAML('tests/sauce_browsers.yml')
                }
            }
        }
    });

    grunt.registerTask('default', [
        'production'
    ]);

    grunt.registerTask('tests', [
        'babel:visualTests',
        'babel:unitTests',
        'less:visualTests',
        'less:unitTests',
        'modernizr',
        'qunit'
    ]);

    grunt.registerTask('production', [
        'eslint',
        'clean',
        'babel:production',
        'less:production',
        'postcss',
        'uglify',
        'cssmin',
        'tests'
    ]);

    grunt.registerTask('development', [
        'eslint',
        'clean',
        'babel:development',
        'less:development',
        'postcss',
        'uglify',
        'cssmin',
        'tests'
    ]);

    let ciTasks = ['production'];

    // See https://docs.travis-ci.com/user/pull-requests/#Security-Restrictions-when-testing-Pull-Requests
    if ('false' === process.env.TRAVIS_PULL_REQUEST) {
        ciTasks.push(
            'connect',
            'saucelabs-qunit'
        );
    }

    grunt.registerTask('ci', ciTasks);
};


js/jquery.scrollbox.js
/**
 * jquery.scrollbox.js
 *
 * @version    1.0.0
 * @author     Max Invis1ble
 * @copyright  (c) 2013-2016, Max Invis1ble
 * @license    MIT http://www.opensource.org/licenses/mit-license.php
 */
+function ($, window, document, undefined) {

    'use strict';

    var name = 'scrollbox',
        Scrollbox = function ($element, options) {
            var o = this.options = $.extend({}, $.fn[name].defaults, options);

            this.$element = $element;
            this.$rail = $(o.templates.rail);
            this.$bar = $(o.templates.bar);

            this.init = o.init || this.init;
            this.addListeners = o.addListeners || this.addListeners;
            this.removeListeners = o.removeListeners || this.removeListeners;
            this.scroll = o.scroll || this.scroll;
            this.jump = o.jump || this.jump;
            this.update = o.update || this.update;
            this.destroy = o.destroy || this.destroy;

            this._isOver = this._isBarCaptured = this._isShown = false;
            this._prevY = 0;
            this._isReachTriggered = { top: false, bottom: false };
            this._scrollHeight = undefined;
            this._setScrolledToY();
            this._barTouchId = this._elementTouchId = this._swipeStartY = this._swipeStartedAt = null;

            this._listeners = {};

            this.init();
        };

    Scrollbox.prototype = {

        init: function () {
            var options = this.options;

            this.$wrapper = this.$element
                .trigger('init.' + name)
                .css('overflow', 'hidden')
                .wrap(options.templates.wrapper).parent()
                .append(this.$rail)
                .append(this.$bar);

            this._updateBarHeight();

            if ('top' !== options.start) {
                this.jump(options.start);
            }

            if (this._isShown) {
                this.addListeners();
            }
        },

        addListeners: function () {
            this._listeners.wheel = $.proxy(this._onWheel, this);

            this.$wrapper.on({
                mouseenter: $.proxy(this._onEnter, this),
                mouseleave: $.proxy(this._onLeave, this)
            });

            if (window.addEventListener) {
                window.addEventListener('DOMMouseScroll', this._listeners.wheel, false);
                window.addEventListener('mousewheel', this._listeners.wheel, false);
            } else {
                document.attachEvent('onmousewheel', this._listeners.wheel);
            }

            this.$bar.on({
                mousedown: $.proxy(this._onBarMouseDown, this),
                touchstart: $.proxy(this._onBarTouchStart, this)
            });

            this.$element.on({
                scroll: $.proxy(this._onElementScroll, this),
                touchstart: $.proxy(this._onElementTouchStart, this)
            });

            $(document).on({
                mouseup: $.proxy(this._onDocumentMouseUp, this),
                mousemove: $.proxy(this._onDocumentMouseMove, this),
                touchend: $.proxy(this._onDocumentTouchEnd, this),
                touchmove: $.proxy(this._onDocumentTouchMove, this)
            });
        },

        removeListeners: function () {
            this.$wrapper.off({
                mouseenter: this._onEnter,
                mouseleave: this._onLeave
            });

            if (window.removeEventListener) {
                window.removeEventListener('DOMMouseScroll', this._listeners.wheel, false);
                window.removeEventListener('mousewheel', this._listeners.wheel, false);
            } else {
                document.detachEvent('onmousewheel', this._listeners.wheel);
            }

            this.$bar.off({
                mousedown: this._onBarMouseDown,
                touchstart: this._onBarTouchStart
            });

            this.$element.off({
                scroll: this._onElementScroll,
                touchstart: this._onElementTouchStart
            });

            $(document).off({
                mouseup: this._onDocumentMouseUp,
                mousemove: this._onDocumentMouseMove,
                touchend: this._onDocumentTouchEnd,
                touchmove: this._onDocumentTouchMove
            });
        },

        _onEnter: function (e) {
            e.preventDefault();

            this._isOver = true;
        },

        _onLeave: function (e) {
            e.preventDefault();

            this._isOver = false;
        },

        _onWheel: function (e) {
            if (this._isOver) {
                if (undefined === e) {
                    e = window.event;
                }

                if (e.preventDefault) {
                    e.preventDefault();
                } else {
                    e.returnValue = false;
                }

                this.scroll((e.detail ? e.detail / 3 : -e.wheelDelta / 120) * this.options.wheelSensitivity);
            }
        },

        _onBarMouseDown: function (e) {
            if (1 === e.which) {
                e.preventDefault();

                this._dragStart(e.pageY);
            }
        },

        _onDocumentMouseMove: function (e) {
            if (this._isBarCaptured) {
                e.preventDefault();

                this._drag(e.pageY);
            }
        },

        _onDocumentMouseUp: function (e) {
            if (this._isBarCaptured && 1 === e.which) {
                e.preventDefault();

                this._dragStop();
            }
        },

        _onBarTouchStart: function (e) {
            var touches = e.originalEvent.targetTouches;

            if (1 == touches.length) {
                e.preventDefault();

                this._barTouchId = touches[0].identifier;
                this._dragStart(touches[0].pageY);
            }
        },

        _onDocumentTouchMove: function (e) {
            var touches = e.originalEvent.targetTouches,
                i;

            if (this._isBarCaptured) {
                for (i in touches) {
                    if (touches[i].identifier === this._barTouchId) {
                        e.preventDefault();

                        this._drag(touches[i].pageY);
                        break;
                    }
                }
            }

            if (null !== this._elementTouchId) {
                for (i in touches) {
                    if (touches[i].identifier === this._elementTouchId) {
                        e.preventDefault();

                        this._swipe(touches[i].pageY);
                        break;
                    }
                }
            }
        },

        _onDocumentTouchEnd: function (e) {
            var touches = e.originalEvent.changedTouches,
                swipeDuration,
                swipeDistance,
                swipeSpeed,
                offset,
                i;

            if (this._isBarCaptured) {
                for (i in touches) {
                    if (touches[i].identifier === this._barTouchId) {
                        e.preventDefault();

                        this._dragStop();
                        this._barTouchId = null;
                        break;
                    }
                }
            }

            if (null !== this._elementTouchId) {
                for (i in touches) {
                    if (touches[i].identifier === this._elementTouchId) {
                        e.preventDefault();

                        swipeDuration = Date.now() - this._swipeStartedAt;

                        if (swipeDuration <= this.options.momentum.thresholdTime) {
                            swipeDistance = this._swipeStartY - touches[i].pageY;
                            swipeSpeed = Math.abs(swipeDistance / swipeDuration);
                            offset = swipeSpeed * swipeSpeed * 2 * this.options.momentum.acceleration;

                            if (swipeDistance < 0) {
                                offset = -offset;
                            }

                            this.scroll(offset, {
                                duration: swipeSpeed * this.options.momentum.acceleration,
                                easing: 'momentum'
                            });
                        }

                        this._swipeStartY = this._swipeStartedAt = this._elementTouchId = null;
                        break;
                    }
                }
            }
        },

        _onElementScroll: function (e) {
            e.preventDefault();

            this._updateBarPosition();
        },

        _onElementTouchStart: function (e) {
            var touches = e.originalEvent.targetTouches;

            if (1 == touches.length) {
                e.preventDefault();

                if (this.$element.is(':animated')) {
                    this.$element.stop(true, false);
                    this._setScrolledToY(this.$element.scrollTop());
                }

                this._elementTouchId = touches[0].identifier;
                this._swipeStartY = this._prevY = touches[0].pageY;
                this._swipeStartedAt = new Date();
            }
        },

        _dragStart: function (y) {
            this._isBarCaptured = true;
            this._prevY = y;

            this.$bar.addClass(name + '-bar-captured');
        },

        _drag: function (y) {
            var elementHeight = this.$element.outerHeight();

            this.scroll((y - this._prevY) * ((this._getScrollHeight() - elementHeight) / (elementHeight - this.$bar.height())));
            this._prevY = y;
        },

        _dragStop: function () {
            this._isBarCaptured = false;

            this.$bar.removeClass(name + '-bar-captured');
        },

        _swipe: function (y) {
            this.scroll(this._prevY - y);
            this._prevY = y;
        },

        scroll: function (delta, animationOptions) {
            var max = this._getScrollHeight() - this.$element.outerHeight(),
                scrollToY = this._getScrolledToY() + delta,
                options = this.options,
                position;

            this.$element
                .trigger('scroll.' + name)
                .stop(true, false);

            if (scrollToY >= max) {
                scrollToY = max;
            } else if (scrollToY <= 0) {
                scrollToY = 0;
            }

            this._setScrolledToY(scrollToY);

            if (undefined === animationOptions) {
                this.$element.scrollTop(scrollToY);
            } else {
                this.$element.animate({
                    scrollTop: scrollToY
                }, animationOptions);
            }

            if (!this._isReachTriggered.bottom && scrollToY + options.distanceToReach >= max) {
                position = 'bottom';
            } else if (!this._isReachTriggered.top && scrollToY - options.distanceToReach <= 0) {
                position = 'top';
            }

            if (position) {
                this.$element.trigger($.Event('reach.' + name, { position: position }));
                this._isReachTriggered[position] = true;
            }
        },

        jump: function (y, animationOptions) {
            if ('top' === y) {
                y = 0;
            } else if ('bottom' === y) {
                y = this._getScrollHeight() - this.$element.height();
            }

            this.scroll(y - this._getScrolledToY(), animationOptions);
        },

        update: function () {
            var isShown = this._isShown;

            this._scrollHeight = undefined;
            this._isReachTriggered.top = this._isReachTriggered.bottom = false;
            this._updateBarHeight();

            if (this._isShown) {
                this._updateBarPosition();

                if (!isShown) {
                    this.addListeners();
                }
            }
        },

        _getScrolledToY: function () {
            if (undefined === this._scrolledToY) {
                this._scrolledToY = this.$element.scrollTop();
            }

            return this._scrolledToY;
        },

        _setScrolledToY: function (y) {
            this._scrolledToY = y;
        },

        _getScrollHeight: function () {
            // opera bug workaround
            if (undefined === this._scrollHeight) {
                this._scrollHeight = this.$element[0].scrollHeight;
            }

            return this._scrollHeight;
        },

        _updateBarHeight: function () {
            var elementHeight = this.$element.outerHeight(),
                ratio = elementHeight / this._getScrollHeight();

            if (1 !== ratio) {
                this.$bar.height(elementHeight * ratio);

                if (!this._isShown) {
                    this.$bar.addClass(name + '-bar-in');
                    this.$rail.addClass(name + '-rail-in');
                    this._isShown = true;
                }
            } else if (this._isShown) {
                this.removeListeners();
                this.$bar.removeClass(name + '-bar-in');
                this.$rail.removeClass(name + '-rail-in');
                this._isShown = false;
            }
        },

        _updateBarPosition: function () {
            var elementHeight = this.$element.outerHeight();

            this.$bar.css(
                'top',
                (elementHeight - this.$bar.outerHeight()) * (this.$element.scrollTop() / (this._getScrollHeight() - elementHeight))
            );
        },

        destroy: function () {
            this.$wrapper
                .off('.' + name)
                .find('*').off('.' + name);

            this.removeListeners();

            this.$element
                .unwrap()
                .removeData(name);

            this.$rail.remove();
            this.$bar.remove();

            this.$bar = this.$rail = this.$element = this.$wrapper = null;
        }

    };

    if ($.easing.momentum === undefined) {
        $.easing.momentum = function (x, t, b, c, d) {
            var ts = (t /= d) * t,
                tc = ts * t;

            return b + c * (-1 * ts * ts + 4 * tc + -6 * ts + 4 * t);
        };
    }

    $.fn[name] = function (option) {
        var args = Array.prototype.slice.call(arguments, 1);

        return this.each(function () {
            var $this = $(this),
                data = $this.data(name),
                options;

            if ('object' === typeof option) {
                options = option;
            }

            if (!data) {
                $this.data(name, (data = new Scrollbox($this, options)));
            }

            if ('string' === typeof option) {
                data[option].apply(data, args);
            }
        });
    };

    $.fn[name].Constructor = Scrollbox;

    $.fn[name].defaults = {
        distanceToReach: 0,
        wheelSensitivity: 20,
        momentum: {
            acceleration: 1600,
            thresholdTime: 500
        },
        start: 'top',
        templates: {
            bar: '<div class="' + name + '-bar"></div>',
            rail: '<div class="' + name + '-rail"></div>',
            wrapper: '<div class="' + name + '-wrapper"></div>'
        }
    };

}(jQuery, window, document);